File: starters/javascript/grid.js
/**
 * Grid utility functions for AI Code Battle.
 *
 * Provides toroidal distance calculations, neighbor enumeration,
 * and BFS pathfinding on a wrapping grid.
 */

function toroidalManhattan(r1, c1, r2, c2, cols, rows) {
  const dr = Math.min(Math.abs(r1 - r2), rows - Math.abs(r1 - r2));
  const dc = Math.min(Math.abs(c1 - c2), cols - Math.abs(c1 - c2));
  return dr + dc;
}

function toroidalChebyshev(r1, c1, r2, c2, cols, rows) {
  const dr = Math.min(Math.abs(r1 - r2), rows - Math.abs(r1 - r2));
  const dc = Math.min(Math.abs(c1 - c2), cols - Math.abs(c1 - c2));
  return Math.max(dr, dc);
}

function neighbors(row, col, rows, cols) {
  const offsets = [
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1],           [0, 1],
    [1, -1],  [1, 0],  [1, 1],
  ];
  return offsets.map(([dr, dc]) => [
    (row + dr + rows) % rows,
    (col + dc + cols) % cols,
  ]);
}

/**
 * BFS pathfinding on a toroidal grid.
 *
 * @param {[number,number]} start - [row, col]
 * @param {[number,number]} goal - [row, col]
 * @param {function(number,number): boolean} passable - returns true if walkable
 * @param {number} rows
 * @param {number} cols
 * @returns {[number,number][]|null} path from start to goal (excl. start), or null
 */
function bfs(start, goal, passable, rows, cols) {
  const [sr, sc] = start;
  const [gr, gc] = goal;
  if (sr === gr && sc === gc) return [];

  const key = (r, c) => `${r},${c}`;
  const visited = new Set([key(sr, sc)]);
  const queue = [{ r: sr, c: sc, path: [] }];

  while (queue.length > 0) {
    const { r, c, path } = queue.shift();
    for (const [nr, nc] of neighbors(r, c, rows, cols)) {
      const newPath = [...path, [nr, nc]];
      if (nr === gr && nc === gc) return newPath;
      const k = key(nr, nc);
      if (!visited.has(k) && passable(nr, nc)) {
        visited.add(k);
        queue.push({ r: nr, c: nc, path: newPath });
      }
    }
  }
  return null;
}

module.exports = {
  toroidalManhattan,
  toroidalChebyshev,
  neighbors,
  bfs,
};

File: starters/csharp/Grid.cs
// Grid utility functions for AI Code Battle.
//
// Provides toroidal distance calculations, neighbor enumeration,
// and BFS pathfinding on a wrapping grid.

using System.Collections.Generic;

static class Grid
{
    private static readonly (int dr, int dc)[] Offsets =
    {
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1),
    };

    /// Manhattan distance with wrap-around on a toroidal grid.
    public static int ToroidalManhattan(int r1, int c1, int r2, int c2, int rows, int cols)
    {
        int dr = Math.Min(Math.Abs(r1 - r2), rows - Math.Abs(r1 - r2));
        int dc = Math.Min(Math.Abs(c1 - c2), cols - Math.Abs(c1 - c2));
        return dr + dc;
    }

    /// Chebyshev distance with wrap-around on a toroidal grid.
    public static int ToroidalChebyshev(int r1, int c1, int r2, int c2, int rows, int cols)
    {
        int dr = Math.Min(Math.Abs(r1 - r2), rows - Math.Abs(r1 - r2));
        int dc = Math.Min(Math.Abs(c1 - c2), cols - Math.Abs(c1 - c2));
        return Math.Max(dr, dc);
    }

    /// 8-directional neighbors with wrap-around.
    public static Position[] Neighbors(Position p, int rows, int cols)
    {
        var result = new Position[8];
        for (int i = 0; i < Offsets.Length; i++)
        {
            result[i] = new Position
            {
                Row = (p.Row + Offsets[i].dr + rows) % rows,
                Col = (p.Col + Offsets[i].dc + cols) % cols,
            };
        }
        return result;
    }

    /// BFS pathfinding on a toroidal grid.
    /// Returns path (excluding start) or null if unreachable.
    public static List<Position>? Bfs(Position start, Position goal,
        Func<Position, bool> passable, int rows, int cols)
    {
        if (start.Row == goal.Row && start.Col == goal.Col)
            return [];

        var visited = new HashSet<(int, int)> { (start.Row, start.Col) };
        var queue = new Queue<(Position pos, List<Position> path)>();
        queue.Enqueue((start, []));

        while (queue.Count > 0)
        {
            var (cur, path) = queue.Dequeue();
            foreach (var nb in Neighbors(cur, rows, cols))
            {
                var newPath = new List<Position>(path) { nb };
                if (nb.Row == goal.Row && nb.Col == goal.Col)
                    return newPath;

                var key = (nb.Row, nb.Col);
                if (!visited.Contains(key) && passable(nb))
                {
                    visited.Add(key);
                    queue.Enqueue((nb, newPath));
                }
            }
        }
        return null;
    }
}
